Serenity.await

Lakelezz opened the pull request:
This pull request rewrites Serenity to be asynchronous.

As highly demanded (#729), Serenity is finally async.
For a long time, it was nothing but a dream, but now it's reality.

There are still a few things following in this PR, but it's ready to be used, see example 5.

Missing bits:
- [x] `voice`-module;
- [x] `voice`-examples (6, 10);
- [x] scheduling/eventing example (12);
- [x] updating all docs, doc-tests, and code examples in other files;
- [x] `native_tls_backend`;
- [ ] final code refactoring;

To fix:
- [x] Serenity becomes unresponsive, no shard activity, goes offline;


Also, collectors (as mentioned in #786) were added too.

You can test it already!

spring4175 committed:
client: remove gateway, rest re-exports

Remove the following exports from the 'client' module:

- crate::gateway
- crate::http as rest

Long, long, long ago, 'http' was called 'rest', and both 'gateway' and
'rest' were located in the 'client' module. These haven't been here
since 0.2.x (May of 2017!), so really people should've moved on at this
point.

Signed-off-by: Zeyla Hellyer <zeyla@hellyer.dev>

spring4175 committed:
http/client: store inner client in an arc

In the 'http' module's 'Http' struct, store the inner reqwest client in
an Arc. This doesn't have much use right now for serenity itself, but it
does allow users to pass in a reference to their own client when making
a client via 'http::Http::new', avoiding the need to have multiple
clients.

The primary use for this is to ease a slight rework of the
'http::ratelimiting' module.

Signed-off-by: Zeyla Hellyer <zeyla@hellyer.dev>

spring4175 committed:
Rename 'raw' module to 'client' (#613)

The 'http::raw' module was initially intended as a home for oneshot
request functions. As this has now gone away and been replaced by a full
client, the module itself should be renamed to 'client' to match it.

spring4175 committed:
http: rework ratelimiting structure

Rework the ratelimiting structure to avoid an odd hierarchy and to
remove the static variable.

Ratelimiting was done through a freefloating function named 'perform'.
It accepted a reference to the Http client instance and the request to
perform. This is backwards in design: the Http client in this design is
passing itself to the ratelimiting module, which in turn calls back into
the Http passed into it to access route ratelimit information and the
global ratelimiter mutex on the Http client. At that point, you may as
well put the ratelimiting on the Http client itself.

Another poor design choice was to use a 'static mut' for storing the
offset. Although this is currently only set once -- after the first
response ever is received -- it's noted[1] in the source right now that
it may be preferable to mutate it more often in the future.

The first issue fixed is the ratelimiting design: the ratelimiter module
now has a struct named 'ratelimiter' that owns:

- its own global ratelimiting mutex
- an Arc<ReqwestClient> that it shares with the Http client
- the offset stored as an atomic
- a map of routes to their ratelimits
- the bot token

With this design, ratelimiting information is self-contained.

The second issue was the offset. The public interface is the same here,
but now on the 'ratelimiter' struct. That is, the function definition
still returns an 'option<i64>', despite the fact that it's internally
stored as an atomic i64. The value is i64::MAX when no offset has been
calculated yet.

While not perfect still, there have been some efficiency improvements.
The lock for each route is no longer held throughout the duration of
each request to that route. This means that 5 message sends in the same
channel, assuming a limit of 5, will happen more or less simultaneously
if sent all in one go on 5 threads.

A minour breaking change is that the 'ratelimit' struct containing the
'limit', 'remaining', and 'reset' for a route are now private
structfields, exposed through public getters. This probably doesn't
affect any users.

Signed-off-by: Zeyla Hellyer <zeyla@hellyer.dev>

Lakelezz committed:
Use `CacheRwLock` in `CacheAndHttp` (#658)

Use `CacheRwLock` in `CacheAndHttp`.
Replacing the normal `Arc<RwLock<Cache>>` with the
API-friendly `CacheRwLock`.
Implement `CacheHttp` for `CacheAndHttp`.
Also takes `Arc` and `&T` into account.

Lakelezz committed:
Display CI-badge for `next`-branch.

After the `v0.6.x`-branch has been deleted, our badge-link no longer pointed to a valid target-branch.
This changes the target to the `next`-branch.

arqunis committed:
Get rid of the `Arc` implementation for `Framework`

This is technically a breaking change, but I doubt its removal will affect anyone.
There's no meaningful usage of an `Arc<T> where T: Framework` without `Mutex` or `RwLock`-based interior-mutability indirection
as the framework method requires mutable access of `self` (`&mut self`).
This can be hard to achieve with `Arc` because there must be only one owner of the pointer at the time, nullifying its usefulness for shared memory.

In summary, this impl is useless.

arqunis committed:
Fix compilation

arqunis committed:
Apply case-insensitivity on prefixes (#690)

arqunis committed:
Turn off the `owner_privilege` option (#692)

Its default value will now be `false`.

ikkerens committed:
Ensure nullable Discord API values are optional in audit log fields (#691)

mendess committed:
Add the ability to have multiple examples in commands (#707)

arqunis committed:
Turn the function-like group macro to an attribute macro (#712)

arqunis committed:
Utilise Discord's new precision and reset-after headers (#699)

ikkerens committed:
Change AttachmentType to use Cows instead (#718)

This commit actually consists of three changes to make `AttachmentType` significantly more useable and (when done properly) significantly more efficient.

- `AttachmentType` variants that used a tuple before now use a struct with named fields, making the code more self-documenting.
- `AttachmentType` variants that needed a filename now use `String` instead of `&str`, preventing a `.to_string()` call later on and giving the end library user the responsibility (which in turn also prevents any potential lifetime pitfalls)
- `AttachmentType::Bytes` now uses a `std::borrow::Cow` instead of a lifetimed-borrow. When uploading attachments it is not always desirable to pass a reference due to the way people structure their applications, especially as later on when actually sending the attachment the reference is copied by calling `to_vec`. Using a cow annihilates this completely because using `Cow::Owned::into_owned()` is a no-op, thus preventing a potential copy of up to 50MB.

arqunis committed:
Format the list of attribute options with tables (#717)

arqunis committed:
Remove unused import

arqunis committed:
Actually, re-add it back, in its proper place.

arqunis committed:
Impose thread-safety requirements on the event handlers (#719)

And, transition their internal usage to trait objects, thus utilising dynamic dispatch in leu of static dispatch.

arqunis committed:
Do not duplicate client initialisation code for event handlers and the cache update timeout  (#720)

arqunis committed:
Add an option to turn off guild subscriptions (#721)

arqunis committed:
Remedy a typo

arqunis committed:
Interpret doc-comments as the description attribute (#724)

Closes #723

arqunis committed:
Re-add the Client constructor methods, but deprecate them

These were removed in https://github.com/serenity-rs/serenity/commit/56fe660b789ab1ac2aad9d75b4d3add39abf1cf8 as they duplicated code from `new`, and have therefore been replaced by `new_with_extras`.

But the replacement doesn't justify their *whole* removal. Instead, they'll serve as shorthands for using
the new method.

arqunis committed:
Remove redundant `cfg`s for the http feature in the client module (#726)

The client module has the http module defined as its dependency; it can NOT function without it. This makes the absency of the http feature impossible, and permits this commit to remove all assumptions of its vacancy in the client code.

arqunis committed:
Account for checks in the help command (#727)

arqunis committed:
Add documentation to the `check` macro

CarlGroth committed:
Mark the event enum as untagged (#738)

ikkerens committed:
Add new auditlog type enums (#744)

* Added new audit log values

* Remove unused import

LeSeulArtichaut committed:
Fix crash due to Bearer token not having the 'email' scope (#746)

nickelc committed:
Upgrade to reqwest v0.10.0

LeSeulArtichaut committed:
Fix regression of the `CommandGroup` struct (#751)

LeSeulArtichaut committed:
Fix Guild::edit_role_position example
Example now mimics GuildId::edit_role_position

Erk- committed:
Add support for receiving custom statuses.

Signed-off-by: Valdemar Erk <valdemar@erk.io>

arqunis committed:
Abuse the compiler to do type checking for us (#759)

* Add the `static_assertions` crate.

* Refactor validation to rely on the compiler

* Feature-gate the re-export

Commit:
Remove april fools functions

Erk- committed:
Simplify the code to get the socket addr for rustls.

Marks as needing at least url 2.1 as it was first introduced
then.

Signed-off-by: Valdemar Erk <valdemar@erk.io>

Erk- committed:
Fix #763

Signed-off-by: Valdemar Erk <valdemar@erk.io>

arqunis committed:
Silence warning emitted on macOS nightly

arqunis committed:
Bump version

arqunis committed:
Update dependencies and the MSRV to 1.39

Lakelezz committed:
Change `http`-module to utilise async-await.
This changes blocking HTTP-requests to be asynchronous.

Lakelezz committed:
Change `http`-module to utilise async-await.
This changes blocking HTTP-requests to be asynchronous.

Lakelezz committed:
Change `model`-module to utilise async-await.
Models will use `tokio`'s `RwLock` and `Mutex` instead of `parking_lot`.

Lakelezz committed:
Change `model`-module to utilise async-await.
Models will use `tokio`'s `RwLock` and `Mutex` instead of `parking_lot`.

Lakelezz committed:
Change `client`-module to utilise async-await.
The `EventHandler` requires async methods now.
Also, the framework no longer sits behind a `Mutex`. This requires the
framework to be constructed at the same time as the client.
The endeavours are not done yet, the different new-functions
will require more flexibiltity to support a framework as argument.

Lakelezz committed:
Change `client`-module to utilise async-await.
The `EventHandler` requires async methods now.
Also, the framework no longer sits behind a `Mutex`. This requires the
framework to be constructed at the same time as the client.
The endeavours are not done yet, the different new-functions
will require more flexibiltity to support a framework as argument.

Lakelezz committed:
Change `gateway`-module to utilise async-await.

Lakelezz committed:
Change `gateway`-module to utilise async-await.

Lakelezz committed:
Change `cache`-module to utilise async-await.

Lakelezz committed:
Change `cache`-module to utilise async-await.

Lakelezz committed:
Change `utils`-module to utilise async-await.

Lakelezz committed:
Change `utils`-module to utilise async-await.

Lakelezz committed:
Update to support async-await.

Lakelezz committed:
Update to support async-await.

Lakelezz committed:
Change `builder`-module to utilise async-await.

Lakelezz committed:
Change `builder`-module to utilise async-await.

Lakelezz committed:
Change `framework`-module to utilise async-await.
This changes the `command_attr`-crate too.
Hooks are no longer closures and must be their own function and
provided as that.
The proc-macro `hook` can help to transform hook functions into
the format required.
Frameworks must be constructed at the same time as the client or
beforehand.

Lakelezz committed:
Change `framework`-module to utilise async-await.
This changes the `command_attr`-crate too.
Hooks are no longer closures and must be their own function and
provided as that.
The proc-macro `hook` can help to transform hook functions into
the format required.
Frameworks must be constructed at the same time as the client or
beforehand.

Lakelezz committed:
Export `tokio`-locks instead of `parking_lot`.

Lakelezz committed:
Export `tokio`-locks instead of `parking_lot`.

Lakelezz committed:
Change bechmarks to work with async-await.

Lakelezz committed:
Change bechmarks to work with async-await.

Lakelezz committed:
Change examples to utilise asynchronous Serenity.
This excludes example 6, 10, and 12.

Lakelezz committed:
Change examples to utilise asynchronous Serenity.
This excludes example 6, 10, and 12.

Lakelezz committed:
Block on all futures.

Lakelezz committed:
Ensure TLS-backends use correct dependencies.

Scotsguy commented:
I've updated my bot to serenity.await already: https://github.com/Scotsguy/background-cat/tree/async
(Here's the updating commit, as an example: https://github.com/Scotsguy/background-cat/commit/fd14980c0ddfaea02a1e03dc15a99a61da1ef5bb)

Lakelezz committed:
Merge branch 'next' into await

Lakelezz committed:
Fix `Client::send_files` for images.
Merging in new changes broke the code.
The async `Reqwest` `Response` does not implement `Read`, this will
attempt to convert the Body to bytes.

Lakelezz committed:
Enable async-await for `DynamicPrefixHook`.

Lakelezz committed:
Remove `thread::sleep` from `internal` and `http`.
These were overseen in the initial port to utilise async-await.

Lakelezz committed:
Change `voice`-module to utilise async-await.
This includes related feature-gated code in other modules.

Lakelezz committed:
Change `voice`-examples to utilise async-await.

Lakelezz committed:
Fix CI when compiling TLS-backends.

Lakelezz committed:
Add missing cfg-check to disallow one TLS-backend.

Lakelezz committed:
Ensure error cfgs consider TLS-backends properly.

Lakelezz committed:
Fix `Error`'s `impl`s TLS-backends usage.

Lakelezz committed:
Fix building without `cache`-feature.

Lakelezz committed:
Turn framework-dispatch into a task.

Lakelezz committed:
Ensure all hooks are awaited.

nickelc commented:
I've successfully updated my bot. :tada: 
Everything seems to work fine with my async coded i use with the current version.

What i don't understand is that `StandardFramework`'s `Args` is partial async.
```rust
let filter = match args.single::<u32>().await {
    Ok(id) => Id::eq(id),
    Err(_) => Fulltext::eq(args.rest()),
};
```

Lakelezz commented:
> What i don't understand is that `StandardFramework`'s `Args` is partial async.

I'm sorry for this, there is an internal locked state happening inside `Args`, however @acdenisSK will rework `Args` soon to enable even more possibilities, it will help with the API too!


> I've successfully updated my bot. 🎉
> Everything seems to work fine with my async coded i use with the current version.

Very glad to hear this! If there is trouble, let me know!

arqunis committed:
Remove partial asynchronosity from Args

arqunis committed:
Fix example

arqunis committed:
Fix more examples

Lakelezz committed:
Remove `Args`' partial async API.

Lakelezz commented:
The latest changes address the partial async API of the `Args`-struct, this introduced new breaking changes - if you ported your application already.

---

On another note, example 12 will be replaced by an example about event collectors.
An event collector, e.g. a message collector, can await replies from a user you are interacting with.
It's currently being implemented, however, there is already a working test implementation, it's not merged yet, but here is a sneak peek:

```rust
    let messages = msg.author.id
        .await_reply()
        .collect_limit(3)
        .timeout(Duration::from_secs(5))
        .receive_all(&ctx)
        .await;

    if messages.is_empty() {
        if let Err(why) = msg.reply(&ctx.http, "You did not send any messages!").await {
            println!("Error sending message: {:?}", why);
        }
    }
```

It will await up to 3 messages sent by the author of `msg` for up to 5 seconds.

hiratara commented:
I am very much looking forward to Serenity.await 😄

I found some bugs in `dispatch.rs`, and I sent some fixes.

* https://github.com/Lakelezz/serenity/pull/2
* https://github.com/Lakelezz/serenity/pull/4

I'm sorry if you've already checked these PRs. I'd appreciate it if these fixes could be available in the `await` branch.

hiratara committed:
Fix building without `framework`-feature

hiratara committed:
Fix using both raw and normal event handler

If both - the event and the raw event handler - were set,
the former received no events
while the latter received them twice.

Lakelezz commented:
Yes, thank you. I merged them!

Lakelezz committed:
Merge branch 'next' into await

nickelc commented:
I've started to fix doc comment examples, starting with `src/client/*`.

arranf commented:
Is there any way to help with this effort?

ankoh commented on src/model/utils.rs:
```diff
@@ -67,7 +67,7 @@ pub fn deserialize_members<'de, D: Deserializer<'de>>(
 
     for member in vec {
-        let user_id = member.user.read().id;
+        let user_id = futures::executor::block_on(member.user.read()).id;
```
My bot is consistently deadlocking here.
I'm fairly new to Rust, but I'm getting here via `async fn recv_event` of the ShardRunner.
Isn't `block_on` in an asynchronous function considered to be very dangerous?

ankoh replied:
```
tokio::task::block_in_place(|| { futures::executor::block_on(member.user.read()).id })
```
This "fixes" the deadlock but would be a rather ugly hack.
The exact same problem occurs quite a few times whenever you de-/serialize a tokio RWLock and later need the ids for HashMaps.

Lakelezz replied:
Yes, this is not good, indeed. I will look into this.

Lakelezz replied:
Here is an insight into the issue: https://github.com/tokio-rs/tokio/issues/2376

ankoh replied:
The idiomatic approach would certainly be an async serde-rs/json that doesnt exist (yet).
https://github.com/serde-rs/json/issues/575

The only option that I currently see is to skip the RWLocks during de-/serialization und introduce them afterwards. Either by using a dedicated copy of the types for serde or by using a private field for each rwlock.

ankoh replied:
Of course, `tokio::task::block_in_place` remains an unattractive, yet working solution.
For deserialization, `block_in_place` doesn't sound super evil since the objects haven't been shared with anyone yet.
For serialization, one could just deeply copy the root to make sure the read will never block as well.

Lakelezz commented:
> 
> Is there any way to help with this effort?

Testing is one way to help with the effort! The more testing gets done, the faster I can merge this into Serenity. The last chunk is being added soon: Collectors!

Lakelezz committed:
Adds collectors for messages and reactions.
This introduces a non-default feature `collector`.
Additionally adds methods on some models to conveniently await
messages or reactions.

Lakelezz committed:
Change example 12 to be about collectors.
The original example's purpose can be solved with collectors.

Lakelezz committed:
Merge branch 'await' of https://github.com/Lakelezz/serenity into await

Lakelezz committed:
Add `collector`-feature to example 12.

Lakelezz committed:
Add missing 'a lifetime for `fetch_single_command`.

Lakelezz committed:
Add missing `collector`-cfgs for the shard runner.

Lakelezz committed:
Remove `block_on`s in the benches.

Lakelezz commented:
Collectors are now usable! Please test and report bugs or design suggestions : )
In order to get started, check out the reworked example 12!

I will address some bugs next!

hiratara commented:
Hello,

I've tested the `06-voice` example and noticed the `~play` command hangs the bot. https://github.com/Lakelezz/serenity/pull/7 fixes the problem.

I also suggest the https://github.com/Lakelezz/serenity/pull/6 for minor refactoring. It's not a mandatory change, but I believe that it is a good thing for the project 😄

Elinvynia committed:
Remove support for Groups (#812)

Groups are not supported with bot accounts.

Elinvynia committed:
Add `guild_id` field to `Role` (#790)

nickelc committed:
Update docs for `async-await`.

* Move the example for `Client::new` to the right doc comment and fix it.

* Fix `serenity::c::Client` examples

* Fix `serenity::c::Context` examples

* Fix `serenity::c::b::gateway::*` examples

* Fix `serenity::http::*` examples

* Fix `serenity::builder::*` examples

* Fix `serenity::framework::*` examples

* Fix `serenity::cache` examples

* Fix `serenity::model` examples

* Fix `serenity::gateway` and `utils` examples

hiratara committed:
Implement async voice process invocations.

Turns the processes for youtube-dl and ffmpeg asynchronous.

hiratara committed:
Unify dispatch with or without `framework`-feature.

Lakelezz commented:
A small tip, due to https://github.com/tokio-rs/tokio/issues/2376 your bot may start stalling.
You can bypass this for now by pinning tokio to "0.2.13" while I'm working on a solution for this problem.

nickelc committed:
Turn `MembersIter` into a stream.

Commit:
Add a new Discord permission for voice channels, STREAM (#816)

Allows the user to go live.

Co-authored-by: Mark Nefedov <mark_nefedov@outlook.com>

hiratara committed:
Fix broken IP Discovery in voice UDP connection (#814)

Lakelezz committed:
Merge branch 'next' into await.

Lakelezz committed:
Remove unused dependencies from example 12.

Lakelezz committed:
Merge branch 'next' into await

Lakelezz committed:
Deduplicate trait-impls.

Lakelezz committed:
Await all `with_mut`s.

Lakelezz committed:
Await framework dispatch.

Lakelezz committed:
Remove unused import in test.

Lakelezz committed:
Deduplicate setting `CONTENT_LENGTH`.

Lakelezz committed:
Await `check_msg` in example 6 and 10.

Lakelezz committed:
Add `cache`-cfg for `PartialGuild::role_by_name`'s test.

Lakelezz committed:
Turn `handle_received_udp` async.

Lakelezz committed:
Fix `ffmpeg_optioned`'s test.

Lakelezz committed:
Fix `GuildId::shard_id`'s test.

Lakelezz committed:
Await `ChannelId::say`s in example 6 and 10.

Lakelezz committed:
Properly handle async in `GuildId::shard_id`.

Lakelezz committed:
Do not await `check_msg`.

Lakelezz committed:
Set cfgs for `EditProfile::avatar`'s test.

Lakelezz committed:
Set cfgs for the `data`-field's test on `Client`.

Celti commented:
Is it expected that `AttachementType` keeps using `std::fs::File` and not `tokio::fs::File`?

Lakelezz committed:
Update collectors to be streams.
The exception is for utility methods on models that await only
a single item, these are still futures only.

Lakelezz commented:
> 
> 
> Is it expected that `AttachementType` keeps using `std::fs::File` and not `tokio::fs::File`?

Might make more sense.

---

On another note, collectors have been reworked! They are now streams. Import `use serenity::futures::stream::StreamExt;` or  `use serenity::futures::stream::TryStreamExt;` to put them to use.
This does not affect utility methods on models that await a single item (`await_reply` and `await_reaction`). These will still be futures returning that optional item.
Their plural variant - `await_reactions` and `await_replies` - return a future building the stream, so add your timeout and other things and then simply await the builder to work with the stream.

Lakelezz committed:
Use `tokio::fs::File` instead of `std::fs::File`.

Lakelezz committed:
Fix tests when using `tokio::fs::File`.

Lakelezz commented:
We switched to `tokio::fs::File` now!

Lakelezz committed:
Use `delay_for` instead of `thread::sleep` in example 9.

Lakelezz committed:
Remove unused import `std::io::Read`.

Lakelezz committed:
Correctly handle rebalancing.

Lakelezz committed:
Spawn blocking to perform `block_on`-deserialisation.

Lakelezz committed:
Merge branch 'next' into await.

Lakelezz committed:
Specify type in doc-tests when creating a `Shard`.
Currently, some types cannot be inferred in Rust.

Lakelezz committed:
Fix doc-test not setting the intents to `None`.

Lakelezz committed:
Remove outdated bits from doc-test.

Lakelezz committed:
Remove artefacts from the merge.

Lakelezz commented:
Just like on the `next` branch, gateway intents are now usable!

Lakelezz committed:
Add example 13 to the CI.

Lakelezz committed:
Merge branch 'next' of https://github.com/serenity-rs/serenity into next

Lakelezz committed:
Merge branch 'next' into await

Lakelezz committed:
Instantly return on `Reidentify`.

Lakelezz committed:
Remove dangerous `block_on`s.

Lakelezz commented:
Tokio blocking/stalling/locking has been fixed! Some models have been adjusted as a result, changes should be straight forward though!

arqunis commented on src/model/channel/mod.rs:
```diff
@@ -203,8 +199,8 @@ impl Channel {
     #[inline]
     pub async fn is_nsfw(&self) -> bool {
```
Redundant `async`
```suggestion
    pub fn is_nsfw(&self) -> bool {
```

arqunis commented on src/model/guild/member.rs:
```diff
@@ -225,7 +222,7 @@ impl Member {
     pub async fn display_name(&self) -> Cow<'_, String> {
```
Ditto
```suggestion
    pub fn display_name(&self) -> Cow<'_, String> {
```

Lakelezz committed:
Turn `async fn`s without `await` to `fn`.

Lakelezz committed:
Correct doc wording for `impl_reaction_collector`.
The method `removed` is intended to mirror `added` in its boolean logic.

Lakelezz committed:
Port example 13 to be asynchronous.

Lakelezz committed:
Handle gateway close frames.
If the error is critical, panic.

Lakelezz committed:
Remove double empty line.

Lakelezz committed:
Redesign client to use a builder impl `Future`.

This design allows to create a client builder, chain methods on it,
and then await the builder, returning a client if successful.

Additionally, one can provide an `Arc`'ed framework and keep an
`Arc` now.
This allows to keep the framework for manual dispatch.

Last but not least, event handlers are no longer required for
the framework.

nickelc commented on examples/02_transparent_guild_sharding/src/main.rs:
```diff
@@ -48,7 +48,10 @@ async fn main() {
     let token = env::var("DISCORD_TOKEN")
         .expect("Expected a token in the environment");
-    let mut client = Client::new(&token, Handler).await.expect("Err creating client");
+    let mut client = Client::new(&token, Handler)
```
```suggestion
    let mut client = Client::new(&token)
```

Lakelezz committed:
Merge branch `next`'s `&mut Context` removal.
Also merges delete message reaction API endpoint support.

Lakelezz committed:
Fix warning in reactionn test.

Elinvynia committed:
Fix empty bytes for image URLs.

Lakelezz committed:
Fix framework multi-dispatch.
When using the `RawEventHandler` and `EventHandler` they
called the `dispatch` again and caused multiple dispatches.

nickelc commented on examples/05_command_framework/src/main.rs:
```diff
@@ -113,32 +119,86 @@ If you want more information about a specific command, just pass the command as
+}
+
+#[hook]
+async fn before(ctx: &mut Context, msg: &Message, command_name: &str) -> bool {
```
```suggestion
async fn before(ctx: &Context, msg: &Message, command_name: &str) -> bool {
```

nickelc commented on examples/05_command_framework/src/main.rs:
```diff
@@ -113,32 +119,86 @@ If you want more information about a specific command, just pass the command as
+}
+
+#[hook]
+async fn after(_ctx: &mut Context, _msg: &Message, command_name: &str, command_result: CommandResult) {
```
```suggestion
async fn after(_ctx: &Context, _msg: &Message, command_name: &str, command_result: CommandResult) {
```

nickelc replied:
+the rest of the hooks

Lakelezz committed:
Fix tests.

Lakelezz committed:
Fix more tests caused by screwed up import-cfgs.

Lakelezz committed:
Turn `&ctx` to `ctx` on `reply`.

Lakelezz committed:
Remove `RawEventHandler` from example 5.

Lakelezz committed:
Use specified features instead of default ones.

Lakelezz committed:
Close shard runners in the `ShardQueuer`.
This allows us to remove the last `block_on` and remove
`blocking` and `rt-threaded` from `tokio`.

Lakelezz committed:
Return `GuildNotFound` when `members` cannot find
the guild inside the cache.

Th3-M4jor committed:
Display Sub Commands in the Help System

Lakelezz committed:
Add `ShardQueuer` debug logs.

Lakelezz committed:
Merge `next`-branch reworking feature-gating.

Lakelezz committed:
Fix feature-gating errors.

Lakelezz committed:
Fix example 13 and test feature-gating errors.

Lakelezz committed:
Improve error when no framework is set.

Lakelezz committed:
Wrap `ShardRunnerInfo`'s `runner_tx` into a `ShardMessenger`.

Lakelezz committed:
Remove `Mutex` around `ShardMessenger`.

Lakelezz committed:
Create `ShardMessenger` in `Context::new` without `cache`-feature.

Lakelezz committed:
Change cache to only hand out clones.
This removes the common cache deadlock,
when the user holds a lock and never releases it.

Lakelezz committed:
Remove `dashmap` as it is not used.

Lakelezz committed:
Update docs referring to `CacheRwLock`.

Lakelezz committed:
Merge branch 'next' into await

Lakelezz committed:
Update `name` to use async and the new cache.

Lakelezz committed:
Remove `pub` on inner cache functions.

Lakelezz committed:
Add cache method to clone current user's fields.
It also adds a method for direct ID copying.

Lakelezz committed:
Add missing docs for cache methods.

Lakelezz committed:
Add cache method to get the count of guild channels.

Lakelezz committed:
Add missing doc for `Client::current_user`.

Lakelezz committed:
Implement `AsRef<Cache>` for `Arc<Context>`.

Lakelezz committed:
Add `gateway`-feature to `serenity`.

Lakelezz committed:
Update example 6 and 10 to use the new cache.

Lakelezz committed:
Add cache methods for users and categories.

Lakelezz committed:
Properly use `guild_channel_field`.

Lakelezz committed:
Add missing `gateway`-feature in example 13.

Lakelezz committed:
Fix warnings in Serenity when building example 1 and 13.

Lakelezz committed:
Add missing doc for `Cache::shard_count`.

Lakelezz committed:
Add method to count guilds in the cache.

Lakelezz committed:
Automatically clone inside field selectors.

Lakelezz committed:
Correct info-tip code position.

Lakelezz committed:
Remove incorrect `Iterator` bound on `guild_channel_field`.

Lakelezz committed:
Remove unnecessary `clone()`-call in doc.

Lakelezz committed:
Borrow field in example for `Cache::member_field`.

Lakelezz committed:
Revert automatic cloning for field accessors.

Th3-M4jor committed:
Propagate gateway intents and token errors up to the client.

Lakelezz commented:
I aim to merge this pull request soon, we are still checking if the OOM bug - that affects only certain users - is fixed with the recent changes.

The final breaking change has been merged, it brings a new cache designed to clone items instead of providing them locked. This includes items' fields, all structs will be without locks now.

About the cache redesign: Deadlocks caused by cache locking are a common and hard to debug error, therefore the cache's fields are now private. The single lock around the cache was broken down into multiple locks around each field.
Furthermore, the known methods will perform a clone, if you wish you to avoid cloning full structs, use the `*_field`-methods that have been added.
They take a closure in which you clone the required field.

Oh and of course: If there is anything missing on the cache, feel free to request it!

Lakelezz committed:
Remove unused collector file.

Lakelezz committed:
Add cache method to set maximum messages to cache.

kageurufu commented:
I migrated my toy bot to use async as a test, and got an oom this morning.

```
[2732306.340126] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/user.slice/user-1000.slice/user@1000.service,task=shitbot,pid=1554827,uid=1000                                            
[2732306.340152] Out of memory: Killed process 1554827 (shitbot) total-vm:10989016kB, anon-rss:10686496kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:21012kB oom_score_adj:0
[2732306.784876] oom_reaper: reaped process 1554827 (shitbot), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```

Let me know what I can do to help

arqunis commented:
Closing this in favour of #905 for continuing the async/await development due to inactivity here.
